### PR TITLE
ngcourse-next Flux Architecture with Rx and Immutable

### PR DESCRIPTION
**Pull request by igor-ka**
hey @yuri, I wanted to get the changes in so as to get feedback sooner than later, but it should probably stay as a pull request for now. Otherwise the rest of the app will be refactored to use this approach next week.

The parts that are worth looking at are TasksStore, TasksActions, TaskList and TaskAddComponents.

@e-schultz @SethDavenport @AbdellaToronto @bertrandk @abmohan @winkerVSbecks
@brianolynyk @hristine

### CODE REVIEW

**Comment by yuri** [comment]
I would merge this sooner rather than later, assuming this runs. Squash the commits, though, and align the commit messages with Rangle's guidelines. (Imperative tense, etc.)

**Comment by igor-ka** [comment]
hi @yuri I squashed the commits, changed the message. The code runs, I double checked.

File: README.md [hunk 1]
```diff
@@ -43,6 +43,12 @@ You will then need to build the front end using:
   gulp build
 ```
 
+You will also need to install TypeScript definition files specified in tsd.json at the root of the project
+
+```
+tsd reinstall -s
```
Comment (brianolynyk) [inline README.md]: If you check in the type definition files you won't need this step.

Comment (igor-ka) [inline README.md]: Yeah I will probably check-in the typings as well in later commits

**Comment by yuri** [comment]
:shipit:

**Comment by igor-ka** [comment]
Hey @yuri and the team.
I have refactored the rest of the application using this approach and got a chance to see the details of the approach in practice. There is still work ahead on this with regards to making it better and cleaner, but this is the first version where the whole of the application has been written this way.

Let me know what you think. Feedback is more than appreciated on this.

**Comment by yuri** [comment]
:shipit:

**Pull request merged**